File: slurmemu/src/interrupt_controller.rs
/*
 *
 *  Interrupt controller
 *
 *
 */

const HSYNC_INTERRUPT : u8 = 1;
const VSYNC_INTERRUPT : u8 = 2;
const AUDIO_INTERRUPT : u8 = 3;
const FLASH_INTERRUPT : u8 = 4;
const GPIO_INTERRUPT  : u8 = 5;
const TIMER_INTERRUPT  : u8 = 6;

pub struct InterruptController {

    interrupt_pending : u8,
    interrupt_enabled : u8,

}

impl InterruptController {

    pub fn new() -> Self {
        InterruptController {
            interrupt_pending : 0,
            interrupt_enabled : 0
        }
    }

    pub fn port_op(&mut self, port : u16, val: u16, write : bool) -> u16
    {
        //println!("Interrupt port op {:#01x} {:#01x} {}", port, val, write);

        match port & 0xf {
            0 => { let a : u8 = self.interrupt_enabled; /* read / set enabled */
                    if write { self.interrupt_enabled = val as u8; }
                    return a as u16;
                },
            1 => { if write { /* clear pending */
                        self.interrupt_pending = self.interrupt_pending & (!val as u8);
                    }
            },
            2 => { return self.interrupt_pending as u16},  /* read out pending */
            _ => ()
        }
        return 0;
    }

    pub fn process_irq(&mut self, hs : bool, vs : bool, audio : bool, flash : bool, gpio : bool, timer: bool) -> u8 /* IRQ -> zero means no IRQ raised */ {

        //println!("process_irq: {} {} {} {}", hs, vs, self.interrupt_pending, self.interrupt_enabled);

        if hs {
            self.interrupt_pending |= 1 << (HSYNC_INTERRUPT - 1);
        }
        if vs {
            self.interrupt_pending |= 1 << (VSYNC_INTERRUPT - 1);
        }
        if audio {
            self.interrupt_pending |= 1 << (AUDIO_INTERRUPT - 1);
        }
        if flash {
            self.interrupt_pending |= 1 << (FLASH_INTERRUPT - 1);
        }
        if gpio {
            self.interrupt_pending |= 1 << (GPIO_INTERRUPT - 1);
        }
	if timer {
	    self.interrupt_pending |= 1 << (TIMER_INTERRUPT - 1);
	}

	/*if a > 1
	{
		println!("Double int");
	}*/

        let ints : Vec<u8> = vec![HSYNC_INTERRUPT, VSYNC_INTERRUPT, AUDIO_INTERRUPT, FLASH_INTERRUPT, GPIO_INTERRUPT, TIMER_INTERRUPT];

        for int in ints {
            if (self.interrupt_pending & self.interrupt_enabled) & (1 << (int - 1)) != 0 {
                //println!("Firing IRQ!{}", int);
                return int;
            }
        }
        return 0;
    }
}

File: slurmemu-ng/src/InterruptController.cpp
/* vim: set et ts=4 sw=4: */

/*
	slurmemu-ng : Next-Generation SlURM16 Emulator

InterruptController.cpp: Handle interrupts

License: MIT License

Copyright 2023 J.R.Sharp

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

#include <InterruptController.h>

InterruptController::InterruptController()
{

}

InterruptController::~InterruptController()
{

}

std::uint16_t InterruptController::port_op(std::uint16_t port, bool write, std::uint16_t wr_val)
{

    if (write)
    {
        switch (port & 0xf)
        {
            case 0:
                m_interrupt_enabled = wr_val;
                break;
            case 1:
                m_interrupt_pending = m_interrupt_pending & (!wr_val);
                break;
        }
    }
    else
    {
        switch (port & 0xf)
        {
            case 0:
                return m_interrupt_enabled;
            case 2:
                return m_interrupt_pending;
        }
    }

    return 0;
}

int InterruptController::step(bool flash_int, bool vs, bool hs, bool timer, bool gpio, bool audio)
{
    if (hs)
        m_interrupt_pending |= 1 << (HSYNC_INTERRUPT - 1);

    if (vs)
        m_interrupt_pending |= 1 << (VSYNC_INTERRUPT - 1);

    if (audio)
        m_interrupt_pending |= 1 << (AUDIO_INTERRUPT - 1);

    if (flash_int)
        m_interrupt_pending |= 1 << (FLASH_INTERRUPT - 1);

    if (gpio)
        m_interrupt_pending |= 1 << (GPIO_INTERRUPT - 1);

    if (timer)
        m_interrupt_pending |= 1 << (TIMER_INTERRUPT - 1);

    int arr[NUM_INTERRUPTS] = {HSYNC_INTERRUPT, VSYNC_INTERRUPT, AUDIO_INTERRUPT, FLASH_INTERRUPT, GPIO_INTERRUPT, TIMER_INTERRUPT}; 

    for (int i = 0; i < NUM_INTERRUPTS; i++)
        if ((m_interrupt_pending & m_interrupt_enabled) & (1 << (arr[i] - 1)))
            return arr[i];

    return 0;
} 

